src/rabbit_channel_interceptor.erl

```diff
@@ -14,73 +14,105 @@
-                                [I, rabbit_misc:method_record_type(M),
-                                 rabbit_misc:method_record_type(M2)])
-    end;
-intercept_method(M, _VHost, Is) ->
-    internal_error("More than one interceptor for method: ~p -- ~p",
-                   [rabbit_misc:method_record_type(M), Is]).
-
-%% select the interceptors that apply to intercept_method().
-select(Method)  ->
-    [M || {_, M} <- rabbit_registry:lookup_all(channel_interceptor),
-          code:which(M) =/= non_existing,
-          M:applies_to(Method)].
+init(Ch) ->
+    Mods = [M || {_, M} <- rabbit_registry:lookup_all(channel_interceptor)],
+    check_no_overlap(Mods),
+    to_fn(Mods, Ch).
+
+%% Turn a list of interceptor modules into a single function
+to_fn(Mods, Ch) ->
+    case Mods of
+        [] -> fun(M, C) -> {M, C} end;
+        [Mod] -> to_fn1(Mod, Ch);
+        [Mod|Mods1] ->
+            Fn1 = to_fn1(Mod, Ch),
+            Fn2 = to_fn(Mods1, Ch),
+            fun(M, C) ->
+                {M1, C1} = Fn1(M, C),
+                Fn2(M1, C1)
+            end
+    end.
+
+%% Turn a single interceptor module into a function
+to_fn1(Mod, Ch) ->
+    St = Mod:init(Ch),
+    fun (M, C) ->
+        % this little dance is because Mod might be unloaded at any point
+        case (catch {ok, Mod:intercept(M, C, St)}) of
+            {ok, R = {M2, _}} ->
+                case validate_method(M, M2) of
+                    true -> R;
+                    _ ->
+                        internal_error("Interceptor: ~p expected to return "
+                                            "method: ~p but returned: ~p",
+                                       [Mod, rabbit_misc:method_record_type(M),
+                                        rabbit_misc:method_record_type(M2)])
+                end;
+            {'EXIT', {undef, [{Mod, intercept, _, _} | _]}} -> {M, C}
+        end
+    end.
+
+check_no_overlap(Mods) ->
+    check_no_overlap1([sets:from_list(Mod:applies_to()) || Mod <- Mods]).
+
+%% Check no non-empty pairwise intersection in a list of sets
+check_no_overlap1(Sets) ->
+    lists:foldl(fun(Set, Union) ->
+                    Is = sets:intersection(Set, Union),
+                    case sets:size(Is) of
+                        0 -> ok;
+                        _ ->
+                          internal_error("Interceptor: more than one "
+                                              "module handles ~p~n", [Is])
+                    end,
+                    sets:union(Set, Union)
+                end,
+                sets:new(),
+                Sets),
+    ok.
+
+intercept_in(#'basic.publish'{} = M, C, S) -> intercept_in1(M, C, S);
+intercept_in(#'basic.ack'{} = M, C, S) -> intercept_in1(M, C, S);
+intercept_in(#'basic.nack'{} = M, C, S) -> intercept_in1(M, C, S);
+intercept_in(#'basic.reject'{} = M, C, S) -> intercept_in1(M, C, S);
+intercept_in(#'basic.credit'{} = M, C, S) -> intercept_in1(M, C, S);
```

Comment:
what's the point of all these function clauses if they all end up in `intercept_in1/3`